### PR TITLE
Fixed the connection factories to implement IAsyncDisposable

### PR DESCRIPTION
**Pull request by bentoi**
I also made minor fixes to the OA ActivateAsync/DisposeAsync methods.

### CODE REVIEW

File: csharp/src/Ice/ConnectionFactory.cs [hunk 1]
```diff
@@ -448,21 +448,51 @@ public void Remove(TKey key, TValue value)
...
 
+        public async ValueTask DisposeAsync()
+        {
+            lock (_mutex)
+            {
+                _disposeTask ??= PerformDestroyAsync();
+            }
+            await _disposeTask.ConfigureAwait(false);
+
+            async Task PerformDestroyAsync()
```
Comment (pepone) [inline csharp/src/Ice/ConnectionFactory.cs:474]: same here

File: csharp/src/Ice/ConnectionFactory.cs [hunk 2]
```diff
@@ -29,11 +29,35 @@ internal sealed class OutgoingConnectionFactory : IConnectionManager
...
```
Comment (pepone) [inline csharp/src/Ice/ConnectionFactory.cs:45]: should we reaname this to `PerformDisposeAsync`?

File: csharp/src/Ice/ObjectAdapter.cs [hunk 3]
```diff
@@ -166,25 +160,19 @@ public async ValueTask DisposeAsync()
...
             {
+                // Synchronously Dispose of the incoming connection factories to stop accepting new incoming requests
+                // or connections. This ensures that once DisposeAsync returns, no new requests will be dispatched.
+                Task[] tasks =
+                    _incomingConnectionFactories.Select(factory => factory.DisposeAsync().AsTask()).ToArray();
+
                 // Wait for activation to complete. This is necessary avoid out of order locator updates.
                 if (_activateTask != null)
                 {
                     await _activateTask.ConfigureAwait(false);
```
Comment (pepone) [inline csharp/src/Ice/ObjectAdapter.cs:171]: Should we also ignore exceptions from _activateTask during dispose?

Comment (bentoi) [reply csharp/src/Ice/ObjectAdapter.cs:171]: Definitely, good catch! I'll fix this.

**Pull request merged**